pkg/gadget-service/api/api.proto

```diff
@@ -248,7 +258,49 @@ service BuiltInGadgetManager {
   rpc RunBuiltInGadget(stream BuiltInGadgetControlRequest) returns (stream GadgetEvent) {}
 }
 
+message InstallGadgetInstanceRequest {
+  GadgetInstance gadgetInstance = 1;
+  int32 eventBufferLength = 2;
+}
+
+message InstallGadgetInstanceResponse {
+  int32 result = 1;
+  GadgetInstance gadgetInstance = 2;
+}
+
+message ListGadgetInstancesRequest {}
+
+message GadgetInstance {
```

Comment:
Yes, let's please discuss the k8s strategy already. I was hoping for feedback like this.

Here's my view:

> versioning strategy

The RunRequest version should be enough ([this](https://github.com/inspektor-gadget/inspektor-gadget/pull/2960#discussion_r1741076291) discussion)

> On Kubernetes, do we store one copy per node or one copy on the cluster?

One copy on the cluster, IMHO. New nodes will automatically be able to pick up the instances and start running them instantly.

> How reconciliation works

Once the node knows about the gadget instance, it will run it. Errors will be kept in memory and can be gotten from the client once connecting (so: we're not keeping the state in Kubernetes resources).

> How users can configure the helm chart to deploy IG with predetermined installed gadgets (I would prefer to use a yaml file rather than marshalled protobuf because it is easier to template from a Helm chart).

Gadget instances can be deployed as additional k8s resources. How that works from Helm, idk.
The k8s resources are not protobuf marshaled (nor contain protobuf marshaled contents), but are simple ConfigMaps _or_ (much more likely now) Custom Resources that are very similar to ConfigMaps (e.g. a couple of attributes + a parameter map as map[string]string)

Since ig-k8s and ig as daemon have slight differences in the way they store the gadget instance configuration, we can use the latter as default and add an option to create the ig-k8s directly from that. I will post a PR for the runtime configuration files soon, which will go a bit more in detail.

Reply:
If it is a ConfigMap, it can be mounted in the pod:
```yaml
  containers:
    - name: ig
      image: ghcr.io/inspektor-gadget/ig:latest
      volumeMounts:
      - name: config-gadgets
        mountPath: /etc/ig/gadets
  volumes:
    - name: config-gadgets
      configMap:
        name: config-gadgets
```
Then, we can have a setup where the ig process does not need to know whether it is running in Kubernetes or as a single host. In this case, the ig process cannot modify the configMap from the file, so this might not be used in setups where ig needs to modify it. But it can still be useful to build custom container images with ig that users can run with the `kubectl debug node` command or with `docker run`.

Custom Resources have the downside that it's difficult to have several installations of Inspektor Gadget in different Kubernetes namespaces (`kubectl gadget deploy --gadget-namespace`). See also https://github.com/inspektor-gadget/inspektor-gadget/pull/2208#issuecomment-2004430802.

Reply:
Just to clarify: it wouldn't be a single Kubernetes resource, but one per gadget instance.

I would have loved to use ConfigMaps to avoid the namespace issue, however:
* we now already use a ConfigMap to carry the IG configuration, so it clashes
* as the CustomResource would have (almost) the same layout as a ConfigMap, it shouldn't need changes in the future; so using one CRD for all versions of IG should be fine™

Using multiple resources (one per instance) can keep the configuration simple (attributes + map[string]string), easy to deploy (no merging / reformatting necessary) and lets you fully use RBAC.

Reply:
> we now already use a ConfigMap to carry the IG configuration, so it clashes

How does it clash? A pod could use several ConfigMaps. The ConfigMaps could use a different name, a different prefix, or different annotations/labels, so it does not need to clash.

> as the CustomResource would have (almost) the same layout as a ConfigMap, it shouldn't need changes in the future; so using one CRD for all versions of IG should be fine™

But it still make it more difficult to deploy several instances of Inspektor Gadget in different namespaces because the CRD is not namespace, so the Helm installer or `kubectl-gadget deploy` installer needs to have some custom reference counting logic.

> Using multiple resources (one per instance) can keep the configuration simple (attributes + map[string]string),

Do you mean one per gadget instance or one per IG deployment instance?

> easy to deploy (no merging / reformatting necessary)

I don't understand why ConfigMap vs CR would matter for merging / reformatting. Which merging / reformatting are you referring to?

> and lets you fully use RBAC.

We can define a RBAC role with a specific `resourceNames`:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/

So you can have different permissions for different ConfigMaps in the same namespace.

It has some limitations (https://github.com/kubernetes/kubernetes/issues/56582) but I am not sure it tips the balance in favour of CRs.

Reply:
> > we now already use a ConfigMap to carry the IG configuration, so it clashes
> 
> How does it clash? A pod could use several ConfigMaps. The ConfigMaps could use a different name, a different prefix, or different annotations/labels, so it does not need to clash.

It (IMHO) clashes in regards to RBAC, see below.

> > as the CustomResource would have (almost) the same layout as a ConfigMap, it shouldn't need changes in the future; so using one CRD for all versions of IG should be fine™
> 
> But it still make it more difficult to deploy several instances of Inspektor Gadget in different namespaces because the CRD is not namespace, so the Helm installer or `kubectl-gadget deploy` installer needs to have some custom reference counting logic.

Agree, that is the downside that we'd have to live with (but this is the case also for other projects, right? How do they handle that case?)

> > Using multiple resources (one per instance) can keep the configuration simple (attributes + map[string]string),
> 
> Do you mean one per gadget instance or one per IG deployment instance?

Per gadget instance.

> > easy to deploy (no merging / reformatting necessary)
> 
> I don't understand why ConfigMap vs CR would matter for merging / reformatting. Which merging / reformatting are you referring to?

That was referring to one ConfigMap vs multiple ConfigMaps (same for one CR vs multiple CR) for the gadget instances - not ConfigMap vs CR.

> > and lets you fully use RBAC.
> 
> We can define a RBAC role with a specific `resourceNames`: https://kubernetes.io/docs/reference/access-authn-authz/rbac/

This is where it clashes. AFAIK (I'll re-check) for resourceNames, no wildcards are allowed. So we cannot use prefixes or such to differentiate access between sub-types of ConfigMaps. Also, in the future we _might_ need yet another type for the "installed gadgets" concept (preloaded, bundled, ...), so doing a catch-all RBAC config for the instance ConfigMaps and then a specific one just for the actual IG configuration ConfigMap (if at all possible) wouldn't be future proof.

> So you can have different permissions for different ConfigMaps in the same namespace.
> 
> It has some limitations ([kubernetes/kubernetes#56582](https://github.com/kubernetes/kubernetes/issues/56582)) but I am not sure it tips the balance in favour of CRs.

Ah, should've read (and clicked the links) to the end first 😄 This is exactly the limitation that applies to us, sadly.

Reply:
> Using multiple resources (one per [gadget] instance) can keep the configuration simple (attributes + map[string]string),

If it is stored in yaml, it could store multiple documents separated by `---`. Then you can have different RBAC rules for the ConfigMap carrying the IG configuration and the ConfigMap carrying the gadgets to start. I don't think we need different RBAC rules for different gadget instances when they are managed by the same ig process anyway.